dynamiq/utils/json_parser.py

```diff
@@ -0,0 +1,86 @@
+import json
+import re
+from typing import Any
+
+
+def extract_json_string(s: str) -> str | None:
+    """Extract the first JSON object from the string by balancing braces.
+
+    Args:
+        s (str): The input string containing the JSON object.
+
+    Returns:
+        Optional[str]: The extracted JSON string, or None if not found.
+    """
+    nesting = 0
+    start = None
+    for i, char in enumerate(s):
+        if char == "{":
```

Comment:
It seems here you handle just object `"{}"` and don't handle list `"[]"`, `"null"` and few other types (int, float, etc)

Reply:
Could you explain that one more time, please?

This function extracts everything that starts with '{' until it finds the closing '}'. So, for instance, if the response looks like this:

```
Here is a JSON
{ "questions": ["Question 1", "Question 2"], 
  "answer": None,
  "contexts": [{"text": "text"}]
}
```

This function will extract:

```
{ "questions": ["Question 1", "Question 2"], 
  "answer": None,
  "contexts": [{"text": "text"}]
}
```

Then, the following function will clean it and attempt to load it as JSON.

Or do you mean we need to extract JSON that is a list of dictionaries as well? Kind of like starting from a list?
```
[{"question": "text"}, {"question": "text 2"}]
```
<img width="483" alt="image" src="https://github.com/user-attachments/assets/98448360-6f58-4d4a-9f88-8161f0320c76" />

Reply:
<img width="540" alt="image" src="https://github.com/user-attachments/assets/033194f8-2d72-4b86-bb26-969ecd71af36" />